### PR TITLE
Checkproofofwork fix

### PR DESCRIPTION
**Pull request by RSerhii**

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -261,10 +261,10 @@ class CTestNetParams : public CChainParams {
         consensus.vDeployments[Consensus::DEPLOYMENT_ASSETS].bit = 5;
         consensus.vDeployments[Consensus::DEPLOYMENT_ASSETS].nStartTime = 1533924000; // GMT: Friday, August 10, 2018 6:00:00 PM
         consensus.vDeployments[Consensus::DEPLOYMENT_ASSETS].nTimeout = 1538351999; // GMT: Sunday, September 30, 2018 11:59:59 PM
-        consensus.vDeployments[Consensus::DEPLOYMENT_SCRYPT2].bit = 7;
-        consensus.vDeployments[Consensus::DEPLOYMENT_SCRYPT2].nStartTime = 1599037200;
-        consensus.vDeployments[Consensus::DEPLOYMENT_SCRYPT2].nTimeout = 1599073200;
-
+        
+        consensus.nNetworkPeriod[Consensus::NETWORK_PERIOD_X16R] = 0;
```
Comment (Krestol) [inline src/chainparams.cpp:265]: I suppose testnet params should be the same as mainnet params

File: src/chainparams.cpp [hunk 2]
```diff
@@ -619,3 +620,32 @@ void TurnOffBIP65() {
 void TurnOffBIP66() {
 	globalChainParams->TurnOffBIP66();
 }
+
+
+Consensus::NetworkPeriod GetNetworkPeriodForBlock(const Consensus::Params& params, int nBlockHeight)
+{
+    int period = 0;
+    for (; period < Consensus::MAX_NETWORK_PERIOD; ++period)
```
Comment (Krestol) [inline src/chainparams.cpp]: it would be better in my opinion

`for (Consensus::NetworkPeriod p = Consensus::MAX_NETWORK_PERIOD; p > NETWORK_PERIOD_X16R; --p)
{
    if (nBlockHeight >= params.nNetworkPeriod[p])
    {
         return p;
    }
}

return NETWORK_PERIOD_X16R;`

File: src/consensus/params.h [hunk 3]
```diff
@@ -17,13 +17,21 @@ enum DeploymentPos
...
     MAX_VERSION_BITS_DEPLOYMENTS
 };
 
+enum NetworkPeriod : int
+{
+    NETWORK_PERIOD_X16R = 0,
+    NETWORK_PERIOD_MAINTANCE,
+    NETWORK_PERIOD_SCRYPT2,
+
+    MAX_NETWORK_PERIOD
```
Comment (Krestol) [inline src/consensus/params.h]: I propose to rename with CURRENT_NETWORK_PERIOD

NETWORK_PERIOD_X16R = 0,
...

CURRENT_NETWORK_PERIOD = NETWORK_PERIOD_SCRYPT2

File: src/primitives/block.cpp [hunk 4]
```diff
@@ -11,58 +11,28 @@
 #include "utilstrencodings.h"
 #include "crypto/common.h"
 #include "crypto/scrypt.h"
+#include "chainparams.h"
```
Comment (Krestol) [inline src/primitives/block.cpp]: I'm not sure it is a good idea  to add dependence from chainparams

File: src/rpc/mining.cpp [hunk 5]
```diff
@@ -146,7 +145,7 @@ UniValue generateBlocks(std::shared_ptr<CReserveScript> coinbaseScript, int nGen
         if (!ProcessNewBlock(Params(), shared_pblock, true, nullptr))
             throw JSONRPCError(RPC_INTERNAL_ERROR, "ProcessNewBlock, block not accepted");
         ++nHeight;
-        blockHashes.push_back(pblock->GetMinedHash(nHeight).GetHex());
+        blockHashes.push_back(pblock->GetMinedHash(Params().GetConsensus(), nHeight).GetHex());
```
Comment (Krestol) [inline src/rpc/mining.cpp]: I suppose it should be GetHash here

File: src/rpc/mining.cpp [hunk 6]
```diff
@@ -132,7 +131,7 @@ UniValue generateBlocks(std::shared_ptr<CReserveScript> coinbaseScript, int nGen
             LOCK(cs_main);
             IncrementExtraNonce(pblock, chainActive.Tip(), nExtraNonce);
         }
-        while (nMaxTries > 0 && pblock->nNonce < nInnerLoopCount && !CheckProofOfWork(pblock->GetMinedHash(nHeight + 1), pblock->nBits, Params().GetConsensus())) {
+        while (nMaxTries > 0 && pblock->nNonce < nInnerLoopCount && !CheckProofOfWork(pblock->GetMinedHash(Params().GetConsensus(), nHeight + 1), pblock->nBits, Params().GetConsensus())) {
```
Comment (Krestol) [inline src/rpc/mining.cpp]: Let's move condition based on chain height to the CheckProofOfWork function
CheckProofOfWork(block, height)
{
 if (Scrypt2Period(height))
 {
 ... 
 }
...
}

File: src/net_processing.cpp [hunk 7]
```diff
@@ -2175,7 +2175,7 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                 // We requested this block for some reason, but our mempool will probably be useless
                 // so we just grab the block via normal getdata
                 std::vector<CInv> vInv(1);
-                vInv[0] = CInv(MSG_BLOCK | GetFetchFlags(pfrom), cmpctblock.header.GetMinedHash(pindex->nHeight));
+                vInv[0] = CInv(MSG_BLOCK | GetFetchFlags(pfrom), cmpctblock.header.GetMinedHash(chainparams.GetConsensus(), pindex->nHeight));
```
Comment (Krestol) [inline src/net_processing.cpp]: GetHash() ?
https://github.com/VeriumReserve/verium/blob/103ecea2e04a64f2a8782aed738f3629a2eb0806/src/net_processing.cpp#L2728
and below

File: src/validation.cpp [hunk 8]
```diff
@@ -3629,25 +3621,32 @@ static bool FindUndoPos(CValidationState &state, int nFile, CDiskBlockPos &pos,
...
+    {
+        if (IsPeriodX16R(consensusParams, pindexPrev->nHeight + 1))
+        {
+            if (!CheckProofOfWork(block.GetHash(), block.nBits, consensusParams)) {
+                return state.DoS(50, false, REJECT_INVALID, "high-hash", false, block.GetHash().ToString() + " proof of work failed ");
+            }
+        }
+        else if (IsPeriodScrypt2(consensusParams, pindexPrev->nHeight + 1))
+        {
+            if (block.nVersion < VERSIONBITS_TOP_BITS_SCRYPT_2) {
```
Comment (Krestol) [inline src/validation.cpp:3662]: Is this check related to CheckProofOfWork?

File: src/validation.cpp [hunk 9]
```diff
@@ -3629,25 +3621,32 @@ static bool FindUndoPos(CValidationState &state, int nFile, CDiskBlockPos &pos,
...
 
-static bool CheckBlockHeaderWorkHash(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, bool fCheckPOW = true)
+static bool CheckWorkBlockHeader(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, bool fCheckPOW = true)
 {
-    // Check proof of work matches claimed amount
-    if (!IsBlockchainX16R() && fCheckPOW && 
-        (!CheckProofOfWork(block.GetNextMinedHash(), block.nBits, consensusParams) ||
-         !CheckProofOfWork(block.GetWorkHash(), block.nBits, consensusParams) )) 
-        return state.DoS(50, false, REJECT_INVALID, "high-hash", false, block.GetNextMinedHash().ToString() + " proof of work failed " + block.GetWorkHash().ToString());
+    CBlockIndex* pindexPrev = chainActive.Tip();
```
Comment (Krestol) [inline src/validation.cpp]: I propose to use height of block.previousBlockHash instead of chainActive.Tip().height

File: src/validation.cpp [hunk 10]
```diff
@@ -3629,25 +3621,32 @@ static bool FindUndoPos(CValidationState &state, int nFile, CDiskBlockPos &pos,
...
-     * once the full data set is received.
-     *
-     */
-
+    // Check header only on recieving blocks
     return true;
 }
 
-static bool CheckBlockHeaderWorkHash(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, bool fCheckPOW = true)
+static bool CheckWorkBlockHeader(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, bool fCheckPOW = true)
```
Comment (Krestol) [inline src/validation.cpp]: 'header' in the name of this function confuses a little bit

File: src/validation.cpp (hunk 9 above)
Comment (Krestol) [inline src/validation.cpp]: if pindexPrev is null CheckWorkBlockHeader returns true is it expected bihevior

File: src/chainparams.cpp (hunk 1 above)
Comment (RSerhii) [reply src/chainparams.cpp:265]: testnet doesn't exist. Periods are set to begin earlier, so if someone will run the testnet he will reach the latest period faster.

File: src/validation.cpp (hunk 8 above)
Comment (RSerhii) [reply src/validation.cpp:3662]: It is related to header validation. It will ensure the creation of hard fork and will force users to update wallets.

File: src/chainparams.cpp (hunk 2 above)
Comment (RSerhii) [reply src/chainparams.cpp]: \+

File: src/consensus/params.h (hunk 3 above)
Comment (RSerhii) [reply src/consensus/params.h]: \+

File: src/primitives/block.cpp (hunk 4 above)
Comment (RSerhii) [reply src/primitives/block.cpp]: \+

File: src/rpc/mining.cpp (hunk 5 above)
Comment (RSerhii) [reply src/rpc/mining.cpp]: \+

File: src/rpc/mining.cpp (hunk 6 above)
Comment (RSerhii) [reply src/rpc/mining.cpp]: \+

File: src/net_processing.cpp (hunk 7 above)
Comment (RSerhii) [reply src/net_processing.cpp]: \+

File: src/validation.cpp (hunk 9 above)
Comment (RSerhii) [reply src/validation.cpp]: \+

File: src/validation.cpp (hunk 10 above)
Comment (RSerhii) [reply src/validation.cpp]: block header is validated in the function. Changed name to ValidateBlockHeader

File: src/validation.cpp (hunk 9 above)
Comment (RSerhii) [reply src/validation.cpp]: fixed

File: src/chainparams.cpp (hunk 1 above)
Comment (Krestol) [reply src/chainparams.cpp:265]: ok

File: src/validation.cpp [hunk 11]
```diff
@@ -3629,25 +3621,53 @@ static bool FindUndoPos(CValidationState &state, int nFile, CDiskBlockPos &pos,
...
-     * once the full data set is received.
-     *
-     */
-
+    // Check header only on recieving blocks
     return true;
 }
 
-static bool CheckBlockHeaderWorkHash(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, bool fCheckPOW = true)
+int GetBlockHeightFromIndex(const CBlockHeader& block)
```
Comment (Krestol) [inline src/validation.cpp:3628]: GetNextBlockHeightFromIndex ?

File: src/validation.cpp [hunk 12]
```diff
@@ -3629,25 +3621,53 @@ static bool FindUndoPos(CValidationState &state, int nFile, CDiskBlockPos &pos,
...
-        return state.DoS(50, false, REJECT_INVALID, "high-hash", false, block.GetNextMinedHash().ToString() + " proof of work failed " + block.GetWorkHash().ToString());
+    if (block.hashPrevBlock == uint256())
+    {
+        return 0;
+    }
+
+    BlockMap::iterator mi = mapBlockIndex.find(block.hashPrevBlock);
+    if (mi == mapBlockIndex.end() || mi->second == nullptr)
+    {
+        return -1;
```
Comment (Krestol) [inline src/validation.cpp:3638]: Add log message

File: src/validation.cpp [hunk 13]
```diff
@@ -3629,25 +3621,53 @@ static bool FindUndoPos(CValidationState &state, int nFile, CDiskBlockPos &pos,
...
+                return state.DoS(50, false, REJECT_INVALID, "high-hash", false, block.GetHash().ToString() + " proof of work failed ");
+            }
+        }
+        else if (IsPeriodScrypt2(consensusParams, height))
+        {
+            if (block.nVersion < VERSIONBITS_TOP_BITS_SCRYPT_2) {
+                return state.DoS(50, false, REJECT_INVALID, "block-version", false, block.GetHash().ToString() + " block version failed ");
+            }
+            if (!CheckProofOfWork(block.GetWorkHash(), block.nBits, consensusParams))  {
+                return state.DoS(50, false, REJECT_INVALID, "high-hash", false, block.GetHash().ToString() + " proof of work failed ");
```
Comment (Krestol) [inline src/validation.cpp:3666]: It would be useful to have network period name in the message. Something like "scrypt^2 proof of work failed"

File: src/validation.cpp (hunk 12 above)
Comment (RSerhii) [reply src/validation.cpp:3638]: in caller function

File: src/validation.cpp (hunk 11 above)
Comment (RSerhii) [reply src/validation.cpp:3628]: we are getting the height of the block passed to the function

**Review by Krestol — APPROVED** [review]

File: src/validation.cpp (hunk 12 above)
Comment (Krestol) [reply src/validation.cpp:3638]: '+

File: src/validation.cpp (hunk 13 above)
Comment (RSerhii) [reply src/validation.cpp:3666]: will be added with autotests
